### PR TITLE
[Prototype] Enable capturing PL objects with JaxPR

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-60576]

### CODE REVIEW

File: pennylane/capture/meta_type.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        "__eq__": lambda self, other: isinstance(other, type(self)),
+        "__hash__": lambda self: hash(type_name),
+    }
+
+    AbstractType = type(type_name, (jax.core.AbstractValue,), namespace)
+
+    # hopefully this API stays constant over jax versions... fingers crossed
+    jax.core.raise_to_shaped_mappings[AbstractType] = lambda aval, _: aval
+
+    ABSTRACT_TYPE_CACHE[type_name] = AbstractType
```
Comment (dwierichs) [inline pennylane/capture/meta_type.py:79]: In order to actually use the caching:
```suggestion
    ABSTRACT_TYPE_CACHE[top_parent] = AbstractType
```

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -838,6 +839,47 @@ def execute_and_compute_vjp(
 
         return tuple(zip(*results))
 
+    def execute_jaxpr(self, jaxpr, *args):
+        """Execute jaxpr using python based simulation utilities.
+
+        Note that at this point, the device *must* pre-specify how many available wires it has.
+        And all the wires must equate to integers starting from zero.
```
Comment (dwierichs) [inline pennylane/devices/default_qubit.py:846]: ?
```suggestion
        And all the wires must equate to consecutive integers starting from zero.
```

File: pennylane/devices/default_qubit.py [hunk 3]
```diff
@@ -47,6 +47,7 @@
 from .execution_config import ExecutionConfig, DefaultExecutionConfig
 from .qubit.simulate import simulate, get_final_state, measure_final_state
 from .qubit.adjoint_jacobian import adjoint_jacobian, adjoint_vjp, adjoint_jvp
+from .qubit.simulate_jaxpr import simulate_jaxpr
```
Comment (dwierichs) [inline pennylane/devices/default_qubit.py:50]: Could go for `simulate_plexpr` :)

File: pennylane/devices/qubit/simulate_jaxpr.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
```
Comment (dwierichs) [inline pennylane/devices/qubit/simulate_jaxpr.py:1]:
```suggestion
# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
```

File: pennylane/capture/meta_type.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
```
Comment (dwierichs) [inline pennylane/capture/meta_type.py:1]:
```suggestion
# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
```

**Comment by trbromley** [comment]
@albi3ro shall we close this prototype branch now that the capture work is well underway?

**Comment by albi3ro** [comment]
I guess I can close this now...

**Pull request closed without merging**